modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -40,108 +41,180 @@
+     * metadata or scripting. That will be handled by copyMetadata and
+     * apply functions that can be overridden.
+     */
+    protected abstract RequestWrapper<?> buildRequest(SearchHit doc);
+
+    /**
+     * Apply various changes depending of the Reindex action and  the search hit,
+     * from copying search hit metadata (parent, routing, etc) to potentially
+     * transforming the {@link RequestWrapper} completely.
+     */
+    protected RequestWrapper<?> apply(RequestWrapper<?> request, SearchHit doc) {
+        return scriptApplier.apply(copyMetadata(request, doc), doc);
+    }
+
+    /**
+     * Copies the metadata from a hit to the request.
+     */
+    protected RequestWrapper<?> copyMetadata(RequestWrapper<?> request, SearchHit doc) {
+        copyParent(request, fieldValue(doc, ParentFieldMapper.NAME));
+        copyRouting(request, fieldValue(doc, RoutingFieldMapper.NAME));
+        copyTimestamp(request, (Long) fieldValue(doc, TimestampFieldMapper.NAME));
+        copyTtl(request, fieldValue(doc, TTLFieldMapper.NAME));
+        return request;
+    }
+
+    /**
+     * Copy the index from a search hit to the request.
+     */
+    protected void copyIndex(RequestWrapper<?> request, String index) {
+        request.setIndex(index);
     }
 
     /**
-     * Part of copyMetadata but called out individual for easy overwriting.
+     * Copy the type from a search hit to the request.
      */
-    protected void copyRouting(IndexRequest index, SearchHit doc) {
-        index.routing(fieldValue(doc, RoutingFieldMapper.NAME));
+    protected void copyType(RequestWrapper<?> request, String type) {
+        request.setType(type);
+    }
+
+    /**
+     * Copy the id from a search hit to the request.
+     */
+    protected void copyId(RequestWrapper<?> request, String id) {
+        request.setId(id);
+    }
+
+    /**
+     * Copy the version from a search hit to the request.
+     */
+    protected void copyVersion(RequestWrapper<?> request, long version) {
+        request.setVersion(version);
+    }
+
+
+    /**
+     * Copy the version type from a search hit to the request.
+     */
+    protected void copyVersionType(RequestWrapper<?> request, VersionType versionType) {
+        request.setVersionType(versionType);
+    }
+
+    /**
+     * Copy the parent from a search hit to the request.
+     */
+    protected void copyParent(RequestWrapper<?> request, String parent) {
+        request.setParent(parent);
+    }
+
+    /**
+     * Copy the routing from a search hit to the request.
+     */
+    protected void copyRouting(RequestWrapper<?> request, String routing) {
+        request.setRouting(routing);
+    }
+
+    /**
+     * Copy the timestamp from a search hit to the request
+     * (if supported by this one)
+     */
+    protected void copyTimestamp(RequestWrapper<?> request, Long timestamp) {
+        copyTimestamp(request, timestamp != null ? timestamp.toString() : null);
```

Comment:
Or maybe none of that has to be true because _ttl only supported on the wrapper that wraps Index requests.